source/vibe/inet/webform.d

```diff
@@ -769,59 +752,159 @@ struct MultipartEntityPart {
+	// @property void contentType(string ct) { headers["Content-Type"] = ct; }
 
 	/** Returns any supplementary parameters of the 'Content-Type' header.
 
 		This is a semicolon separated ist of key/value pairs. Usually, if set,
 		this contains the character set used for text based content types.
 	*/
-	@property string contentTypeParameters()
-	const {
-		auto pv = "Content-Type" in headers;
-		if( !pv ) return "charset=US-ASCII";
-		auto idx = std.string.indexOf(*pv, ';');
-		return idx >= 0 ? (*pv)[idx+1 .. $] : null;
-	}
+	// @property string contentTypeParameters()
+	// const {
+	// 	auto pv = "Content-Type" in headers;
+	// 	if( !pv ) return "charset=US-ASCII";
+	// 	auto idx = std.string.indexOf(*pv, ';');
+	// 	return idx >= 0 ? (*pv)[idx+1 .. $] : null;
+	// }
+}
 
-	/// Creates a multipart entity part as a named form item.
-	static MultipartEntityPart ofFormInput(T)(string name, T v)
-	if (__traits(compiles, to!string(T.init))) {
-		auto part = new MultipartEntityPart();
-		part.headers["Content-Disposition"] = "form-data; name=" ~ name;
-		part.entityBody = v.to!string;
-		return part;
-	}
 
-	/// A builder method creating a part by loading a file by its path.
-	static MultipartEntityPart ofFormFile(string name, string filePath) {
-		import std.path : baseName;
-		import vibe.core.file : openFile;
-		string fileName = baseName(filePath);
-		auto fileStream = interfaceProxy!InputStream(openFile(filePath));
-		return MultipartEntityPart.ofFormFile(name, fileName, getMimeTypeForFile(filePath), fileStream);
-	}
+////
+// Factory Methods
+////
+
+// /// Returns a MultipartEntity with Content-Type "multipart/form-data" from parts as a compile-time sequence.
+// auto multipartFormData(MultipartR...)(MultipartR parts) {
+// 	return multipartFormData(parts);
+// }
+
+// ///
+// unittest {
+// 	import vibe.stream.memory;
+
+// 	// Build an entity using the var-args form.
+// 	auto entity = multipartFormData(
+// 			multipartFormInput("name", "Bob Jones"),
+// 			multipartFormFile("resume", formFile("Resume-Bob.pdf", "application/pdf")),
+// 			multipartFormFiles("photos", [
+// 					formFile("portrait1.jpg", createMemoryStream(cast(ubyte[]) "dummy data"), "image/png"),
+// 					formFile("portrait2.jpg", createMemoryStream(cast(ubyte[]) "dummy data"), "image/png"),
+// 					]),
+// 			);
+// }
+
+/// Returns a MultipartEntity with Content-Type "multipart/form-data" from parts as a range.
+auto multipartFormData(MultipartR)(MultipartR parts)
+if (isInputRange!MultipartR && isMultipartBodyType!(MultipartR)) {
+	string boundaryStr = createBoundaryString();
+	auto headers = only(
+			tuple("Content-Type", "multipart/form-data; boundary=" ~ boundary));
+	return MultipartEntity!(typeof(headers), typeof(parts))(headers: headers, bodyValue: parts, boundary: boundaryStr);
+}
+
+///
+unittest {
+	import vibe.stream.memory;
+
+	// Build an entity using the var-args form.
+	auto entity = multipartFormData(only(
```

Comment:
@s-ludwig I'm currently getting the error below here. At the heart of the matter, is that I have a template which satisfies `isInputStream`, and a range of objects, which itself is a template, that holds these things. That templated range of templated objects matching `isInputStream` struggles to have a type for itself.

However, I cannot use SumType, because the stream type itself has no concrete type.

Any thoughts on how to handle this? Should I drop using templates at a particular level? Is there a concrete type I can use for `InputStream` instead? Perhaps `InputStreamProxy`?

```
source/vibe/inet/webform.d(819,38): Error: none of the overloads of template `std.range.only` are callable using argument types `!()(MultipartEntity!(OnlyResult!(Tuple!(string, string)), string), MultipartEntity!(OnlyResult!(Tuple!(string, string), Tuple!(string, string), Tuple!(string, string)), InterfaceProxy!(InputStream)), MultipartEntity!(OnlyResult!(Tuple!(string, string), Tuple!(string, string)), MapResult!(__lambda6, FormFile!(MemoryStream)[])))`
/usr/include/dmd/phobos/std/range/package.d(10340,6):        Candidates are: `only(Values...)(return scope Values values)`
  with `Values = (MultipartEntity!(OnlyResult!(Tuple!(string, string)), string), MultipartEntity!(OnlyResult!(Tuple!(string, string), Tuple!(string, string), Tuple!(string, string)), InterfaceProxy!(InputStream)), MultipartEntity!(OnlyResult!(Tuple!(string, string), Tuple!(string, string)), MapResult!(__lambda6, FormFile!(MemoryStream)[])))`
  must satisfy the following constraint:
 `       !is(CommonType!Values == void)`
 /usr/include/dmd/phobos/std/range/package.d(10347,6):                        `only()()`
```